clippy_lints/src/clamping_without_clamp.rs

```diff
@@ -0,0 +1,766 @@
+    /// # ;
+    /// ```
+    #[clippy::version = "1.65.0"]
+    pub CLAMPING_WITHOUT_CLAMP,
+    complexity,
+    "using a clamp pattern instead of the clamp function"
+}
+declare_lint_pass!(ClampingWithoutClamp => [CLAMPING_WITHOUT_CLAMP]);
+
+type ExprRef<'tcx> = &'tcx Expr<'tcx>;
+
+#[derive(Debug)]
+struct InputMinMax<'tcx> {
+    input: ExprRef<'tcx>,
+    min: ExprRef<'tcx>,
+    max: ExprRef<'tcx>,
+    add_semicolon: bool,
+    span: Span,
+}
+
+impl<'tcx> LateLintPass<'tcx> for ClampingWithoutClamp {
+    fn check_expr(&mut self, ctx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+        if !expr.span.from_expansion() {
+            check_anti_patterns(ctx, expr);
+        }
+    }
+}
+
+fn check_anti_patterns<'tcx>(cx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+    let suggestion_iter = is_if_else_pattern(cx, expr)
+        .or_else(|| is_max_min_pattern(cx, expr))
+        .or_else(|| is_match_pattern(cx, expr))
+        .into_iter()
+        .chain(is_two_if_pattern(cx, expr).into_iter());
+    for InputMinMax {
+        input,
+        min,
+        max,
+        span,
+        add_semicolon,
+    } in suggestion_iter
+    {
+        let input = Sugg::hir(cx, input, "..").maybe_par();
+        let min = Sugg::hir(cx, min, "..");
+        let max = Sugg::hir(cx, max, "..");
+        let semicolon = if add_semicolon { ";" } else { "" };
+        let suggestion = format!("{input}.clamp({min}, {max}){semicolon}");
+        span_lint_and_sugg(
+            cx,
+            CLAMPING_WITHOUT_CLAMP,
+            span,
+            "clamp-like pattern without using clamp function",
+            "replace with clamp",
+            suggestion,
+            Applicability::MachineApplicable,
+        );
+    }
+}
+/// Targets patterns like
+///
+/// ```
+/// # let (input, min, max) = (0, -3, 12);
+///
+/// if input < min {
+///     min
+/// } else if input > max {
+///     max
+/// } else {
+///     input
+/// }
+/// # ;
+/// ```
+#[allow(clippy::too_many_lines)] // Could be refactored but it works and lints are worked on infrequently.
+fn is_if_else_pattern<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> Option<InputMinMax<'tcx>> {
+    if let ExprKind::If(
+        check,
+        then_body,
+        Some(Expr {
+            kind: ExprKind::If(else_if_check, else_if_body, Some(else_expr)),
+            ..
+        }),
+    ) = simplify_kind(expr).kind
+    && let
+    (
+        ExprKind::Binary(
+            BinOp { node: BinOpKind::Le | BinOpKind::Lt | BinOpKind::Ge | BinOpKind::Gt, ..},
+            first_left,
+            first_right
+        ),
+        ExprKind::Binary(
+            BinOp { node: BinOpKind::Le | BinOpKind::Lt | BinOpKind::Ge | BinOpKind::Gt, ..},
+            second_left,
+            second_right
+        ),
+    ) = (&simplify_kind(check).kind, &simplify_kind(else_if_check).kind)
+    && is_clampable(cx, cx.typeck_results().expr_ty(first_left))
+    && !simplify_kind(then_body).can_have_side_effects()
```

Comment:
https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.peel_blocks_with_stmt.html can be used to extract the single expr/statement expr from a block

Reply:
Done!